apps/snmptrap.1
.RC $Header: /nfs/medea/u0/rel5/rcs/Tools/cmusnmp/apps/snmptrap.1,v 1.4 1995/02/16 16:39:05 djw Exp $
.\" /***********************************************************
.\" 	Copyright 1988, 1989 by Carnegie Mellon University
.\" 
.\"                       All Rights Reserved
.\" 
.\" Permission to use, copy, modify, and distribute this software and its 
.\" documentation for any purpose and without fee is hereby granted, 
.\" provided that the above copyright notice appear in all copies and that
.\" both that copyright notice and this permission notice appear in 
.\" supporting documentation, and that the name of CMU not be
.\" used in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.  
.\" 
.\" CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\" ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
.\" CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
.\" ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
.\" ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
.\" SOFTWARE.
.\" ******************************************************************/
.TH SNMPTRAP 1 "17 September 1989"
.UC 4
.SH NAME
snmptrap - Send an SNMP TRAP message to a host.
.SH SYNOPSIS
snmptrap host community trap-type specific-type [ -a agent-addr ] [-e enterprise-oid] [-d] [-r repeat-count#] [-P port]
.SH DESCRIPTION
snmptrap is an SNMP application that forms and sends an SNMP TRAP message to
a host.
.PP
The
.I host
specification may be either a host name or an internet address
specified in "dot notation"
.PP
The
.I community
specifies the community name for the transaction with the remote system.
.PP
The
.I trap-type
and
.I specific-type
are integers that specify the type of trap message being sent.
.PP
The optional argument
.I -a agent-addr
can be used to change the address from which the trap reports it is being 
sent.  Otherwise the sending host\'s address is used.
.PP
The optional argument
.I -e enterprise-oid
can be used to specify the enterprise object id of the device emitting 
the trap.  The default is 1.3.6.1.4.1.3.1.1.
.PP
The optional argument
.I -r repeat-count#
can be used to specify a repeat count to send the specified trap more than
once.  This is very useful for testing trap receivers.  The traps are sent
very close together in time.
.PP
The
.I -P port_number
option sets the UDP port number to which to send the trap.  If not 
present, the standard SNMP value of 162 is used.
.PP
After the command line arguments are read and parsed, the program will
prompt for the entry of a list of variables to send in the
VarBindingList in the TRAP PDU.
It will prompt with:
.PP
.I Please enter the variable name:
.PP
At this point you can enter one or more variable names, one per line.
Type just a blank line at that prompt to send the TRAP PDU and exit.
Each variable name is given in the format specified in
.I variables (5).
The prompt:
.PP
.I Please enter variable type [i|s|x|d|n|o|t|a]: 
.PP
requests the type of the variable be entered.  Type "i" for an
integer, "s" for an displayable string,
and "n" for a null value. (more...)
.br
At this point a value will be prompted for:
.PP
.I Please enter the value: 
.PP

If this is an integer value, just type the integer (in decimal).  If
it is a string, type in the string. (more...)

For example
.PP
.br
.I snmptrap nic.andrew.cmu.edu public 0 0
.br
.I Please enter the variable name: system.sysdescr.0
.br
.I Please enter variable type [i|s|x|d|n|o|t|a]: s
.br
.I Please enter the value: Unix 4.3BSD-Reno
.br
.I Please enter the variable name: 
.PP
Will send a trap with one variable.
.PP
The defined trap types are:
.TP
.B 0
coldStart \- signifies that the sending protocol entity is
reinitializing itself such that the agent\'s configuration or the
protocol entity implementation may be altered.
.TP
.B 1
warmStart \- signifies that the sending protocol entity is
reinitializing itself such that neither the agent configuration nor
the protocol entity implementation is altered.
.TP
.B 2
linkDown \- signifies that the sending protocol entity
recognizes a failure in one of the communication links represented in
the agent\'s configuration.
.TP
.B 3
linkUp \- signifies that the sending protocol entity
recognizes that one of the communication links represented in the
agent\'s configuration has come up.
.TP
.B 4
authenticationFailure \- signifies that the sending protocol
entity is the addressee of a protocol message that is not properly
authenticated.  While implementations of the SNMP must be capable of
generating this trap, they must also be capable of suppressing the
emission of such traps via an implementation-specific mechanism.
.TP
.B 5
egpNeighborLoss \- signifies that an EGP neighbor for whom
the sending protocol entity was an EGP peer has been marked down and
the peer relationship no longer obtains.
.TP
.B 6
enterpriseSpecific \- signifies that the sending protocol
entity recognizes that some enterprise-specific event has occurred.
The specific-trap field identifies the particular trap which
occurred.
.PP
Adding a "-d" to the argument list will cause the application to dump the output packet.
.PP
.SH "SEE ALSO"
variables(5), RFC 1065, RFC 1212, RFC 1213




HISTORY
Version .91:  Changed the ASN type for the types: Counter, Ipaddress, gauge,
timeticks, and opaque.  The SMI defines them as application specific types,
not ASN_INTEGER and ASN_OCTET_STRING.

Version 1.0: Added snmpnetstat, snmptrap, snmptrapd.  Everything runs on a new
API that has an asynchronous interface.  Most old hard-coded limits are either
non-existent or much larger than necessary.  The mib module now gets it's data
by parsing a text file in the same ASN.1 format as is given in RFC1066.

Version 1.1b: <Not documented>

Version 1.2u: Added Concise MIB parsing support to the snmp library
and to the snmptranslate program.  New mib.txt file that is an
almost-direct extract from RFC1213.  Added the snmproute,
snmptranslate, and snmpdelta programs.  Assorted bug fixes.
Integers are now the only signed numbers.  The snmptrap program is
heavily revised-- it now can pass user-specified varbindings in the
trap PDU.  Some nice snmpnetstat additions from Havard Eidnes
(he@idt.unit.no).


apps/snmptrap.c
/*
 * snmptrap.c - send snmp traps to a network entity.
 *
 */
/***********************************************************
	Copyright 1989 by Carnegie Mellon University

                      All Rights Reserved

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose and without fee is hereby granted, 
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in 
supporting documentation, and that the name of CMU not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.  

CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.
******************************************************************/
#include <sys/types.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <ctype.h>
#include <netdb.h>
#include <stdio.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <sys/sockio.h>
#include <sys/file.h>
#include <nlist.h>

#include "snmp.h"
#include "asn1.h"
#include "snmp_impl.h"
#include "snmp_api.h"
#include "snmp_client.h"

extern int  errno;
int	snmp_dump_packet = 0;

#define NUM_NETWORKS	16   /* max number of interfaces to check */

oid objid_enterprise[] = {1, 3, 6, 1, 4, 1, 3, 1, 1};
oid objid_sysdescr[] = {1, 3, 6, 1, 2, 1, 1, 1, 0};

struct nlist nl[] = {
    { "_boottime" },
    { "" }
};


int snmp_input() {
}

#ifndef IFF_LOOPBACK
#define IFF_LOOPBACK 0
#endif
#define LOOPBACK    0x7f000001
u_long
get_myaddr() {
    int sd;
    struct ifconf ifc;
    struct ifreq conf[NUM_NETWORKS], *ifrp, ifreq;
    struct sockaddr_in *in_addr;
    int count;
    int interfaces;		/* number of interfaces returned by ioctl */

    if ((sd = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
	return 0;
    ifc.ifc_len = sizeof(conf);
    ifc.ifc_buf = (caddr_t)conf;
    if (ioctl(sd, SIOCGIFCONF, (char *)&ifc) < 0) {
	close(sd);
	return 0;
    }
    ifrp = ifc.ifc_req;
    interfaces = ifc.ifc_len / sizeof(struct ifreq);
    for(count = 0; count < interfaces; count++, ifrp++) {
	ifreq = *ifrp;
	if (ioctl(sd, SIOCGIFFLAGS, (char *)&ifreq) < 0)
	    continue;
	in_addr = (struct sockaddr_in *)&ifrp->ifr_addr;
	if ((ifreq.ifr_flags & IFF_UP)
	    && (ifreq.ifr_flags & IFF_RUNNING)
	    && !(ifreq.ifr_flags & IFF_LOOPBACK)
	    && in_addr->sin_addr.s_addr != LOOPBACK) {
		close(sd);
		return in_addr->sin_addr.s_addr;
	    }
    }
    close(sd);
    return 0;
}

/*
 * Returns uptime in centiseconds(!).
 */
long uptime() {
    struct timeval boottime, now, diff;
    mode_t kmem;

    if ((kmem = open("/dev/kmem", 0)) < 0)
	return 0;
/*    nlist("/vmunix", nl);*/
    nlist("/dev/ksyms", nl);
    if (nl[0].n_type == 0) {
	close(kmem);
	return 0;
    }
    
    lseek(kmem, (long)nl[0].n_value, L_SET);
    read(kmem, &boottime, sizeof(boottime));
    close(kmem);

    gettimeofday(&now, 0);
    now.tv_sec--;
    now.tv_usec += 1000000L;
    diff.tv_sec = now.tv_sec - boottime.tv_sec;
    diff.tv_usec = now.tv_usec - boottime.tv_usec;
    if (diff.tv_usec > 1000000L) {
	diff.tv_usec -= 1000000L;
	diff.tv_sec++;
    }
    return ((diff.tv_sec * 100) + (diff.tv_usec / 10000));
}

u_long parse_address(address)
    char *address;
{
    u_long addr;
    struct sockaddr_in saddr;
    struct hostent *hp;

    if ((addr = inet_addr(address)) != -1)
	return addr;
    hp = gethostbyname(address);
    if (hp == NULL) {
	fprintf(stderr, "unknown host: %s\n", address);
	return 0;
    } else {
	bcopy((char *)hp->h_addr, (char *)&saddr.sin_addr, hp->h_length);
	return saddr.sin_addr.s_addr;
    }

}
main(argc, argv)
    int	    argc;
    char    *argv[];
{
    struct snmp_session session, *ss;
    struct snmp_pdu *pdu;
    struct variable_list *vars;
    int	arg;
    char *gateway = NULL;
    char *community = NULL;
    char *trap = NULL, *specific = NULL, *agent = NULL;
    int dest_port = SNMP_TRAP_PORT;

    u_char  value[256];
    int	    name_length;	/* number of subid's in name */
    int	    got_enterprise = 0;
    int	    repeat_count = 1;
    struct variable_list    *vp;
    int     ret;

    init_mib();

    for(arg = 1; arg < argc; arg++)
    {
	if (argv[arg][0] == '-')
	{
	    switch(argv[arg][1])
	    {
	      case 'a':
		agent = argv[++arg];
		break;
	      case 'd':
		snmp_dump_packet++;
		break;
	      case 'r':
		if ((arg + 1) < argc)
		    repeat_count = atoi(argv[++arg]);
		else
		    goto usage;
		break;

	    case 'P':
		if (arg == argc - 1)
		{
		    printf("No value given to -P option\n");
		    exit(1);
		}
		dest_port = atoi(argv[++arg]);
		if (dest_port == 0)
		{
		    printf("Bad port number %s\n", argv[arg]);
		    exit(1);
		}
		break;

	      case 'e':
		got_enterprise = TRUE;
		arg++;
		name_length = MAX_NAME_LEN;
		if (!read_objid(argv[arg], value, &name_length))
		    goto usage;
		break;

	      default:
		printf("invalid option: -%c\n", argv[arg][1]);
		goto usage;
	    }
	    continue;
	}
	if (gateway == NULL)
	    gateway = argv[arg];
	else if (community == NULL)
	    community = argv[arg]; 
        else if (trap == NULL)
	    trap = argv[arg];
	else if (specific == NULL)
	    specific = argv[arg];
    }
    
    if (!(gateway && community && trap && specific))
	goto usage;

    bzero((char *)&session, sizeof(struct snmp_session));
    session.peername = gateway;
    session.community = (u_char *)community;
    session.community_len = strlen((char *)community);
    session.retries = SNMP_DEFAULT_RETRIES;
    session.timeout = SNMP_DEFAULT_TIMEOUT;
    session.authenticator = NULL;
    session.callback = snmp_input;
    session.callback_magic = NULL;
    session.remote_port = dest_port;
    ss = snmp_open(&session);
    if (ss == NULL) {
	printf("Couldn't open snmp\n");
	exit(-1);
    }

    pdu = snmp_pdu_create(TRP_REQ_MSG);
    if (got_enterprise)
    {
	pdu->enterprise = (oid *)malloc(name_length * sizeof(oid));
	bcopy((char *)value, (char *)pdu->enterprise,
	      name_length*sizeof(oid));
	pdu->enterprise_length = name_length;
    }
    else
    {
	pdu->enterprise = (oid *)malloc(sizeof(objid_enterprise));
	bcopy((char *)objid_enterprise, (char *)pdu->enterprise,
	      sizeof(objid_enterprise));
	pdu->enterprise_length = sizeof(objid_enterprise) / sizeof(oid);
    }
    if (agent != NULL)
	pdu->agent_addr.sin_addr.s_addr = parse_address(agent);
    else
	pdu->agent_addr.sin_addr.s_addr = get_myaddr();
    pdu->trap_type = atoi(trap);
    pdu->specific_type = atoi(specific);
    pdu->time = uptime();
    
    pdu->variables = vars = NULL;

    for(ret = 1; ret != 0;)
    {
	vp = (struct variable_list *)malloc(sizeof(struct variable_list));
	vp->next_variable = NULL;
	vp->name = NULL;
	vp->val.string = NULL;

	ret = input_variable(vp);
	if (ret == -1)
	    exit(1);

	if (ret == 1)
	{
	    /* add it to the list */
	    if (vars == NULL)
		/* if first variable */
		pdu->variables = vp;
	    else
		vars->next_variable = vp;

	    vars = vp;
	}
	else
	{
	    /* free the last (unused) variable */
	    if (vp->name)
		free((char *)vp->name);
	    if (vp->val.string)
		free((char *)vp->val.string);
	    free((char *)vp);
	}
    }

    for (; repeat_count > 0; repeat_count--)
    {
	if (snmp_send(ss, pdu)== 0)
	{
	    printf("error\n");
	    snmp_close(ss);
	    exit(1);
	}
	
	/*
	 * want to sleep if sending many traps
	 * give the destination a chance to read
	 */
	if ((repeat_count % 8) == 0)
	    sleep(1);	
    }
    snmp_close(ss);

    exit(0);

  usage:
    printf("usage: snmptrap host community trap-type specific-type \n\
\t[ -a agent-addr ] [-e enterprise-oid] [-d] [-r repeat-count#] [-P port#]\n");
    exit(1);
    /*NOTREACHED*/
}



int
ascii_to_binary(cp, bufp)
    u_char  *cp;
    u_char *bufp;
{
    int	subidentifier;
    u_char *bp = bufp;

    for(; *cp != '\0'; cp++){
	if (isspace(*cp))
	    continue;
	if (!isdigit(*cp)){
	    fprintf(stderr, "Input error\n");
	    return -1;
	}
	subidentifier = atoi(cp);
	if (subidentifier > 255){
	    fprintf(stderr, "subidentifier %d is too large ( > 255)\n", subidentifier);
	    return -1;
	}
	*bp++ = (u_char)subidentifier;
	while(isdigit(*cp))
	    cp++;
	cp--;
    }
    return bp - bufp;
}


int
hex_to_binary(cp, bufp)
    u_char  *cp;
    u_char *bufp;
{
    int	subidentifier;
    u_char *bp = bufp;

    for(; *cp != '\0'; cp++){
	if (isspace(*cp))
	    continue;
	if (!isxdigit(*cp)){
	    fprintf(stderr, "Input error\n");
	    return -1;
	}
	sscanf(cp, "%x", &subidentifier);
	if (subidentifier > 255){
	    fprintf(stderr, "subidentifier %d is too large ( > 255)\n", subidentifier);
	    return -1;
	}
	*bp++ = (u_char)subidentifier;
	while(isxdigit(*cp))
	    cp++;
	cp--;
    }
    return bp - bufp;
}


int
input_variable(vp)
    struct variable_list    *vp;
{
    u_char  buf[256], value[256], ch;

    printf("Please enter the variable name: ");
    fflush(stdout);
    if (gets(buf) == NULL)
	exit(0);

    if (*buf == 0)
    {
	vp->name_length = 0;
	return 0;
    }

    vp->name_length = MAX_NAME_LEN;
    if (!read_objid(buf, value, &vp->name_length))
	return -1;
    vp->name = (oid *)malloc(vp->name_length * sizeof(oid));
    bcopy((char *)value, (char *)vp->name, vp->name_length * sizeof(oid));

    printf("Please enter variable type [i|s|x|d|n|o|t|a]: ");
    fflush(stdout);
    if (gets(buf) == NULL)
	exit(0);
    ch = *buf;
    switch(ch)
    {
      case 'i':
	vp->type = INTEGER;
	break;
      case 's':
	vp->type = STRING;
	break;
      case 'x':
	vp->type = STRING;
	break;
      case 'd':
	vp->type = STRING;
	break;
      case 'n':
	vp->type = NULLOBJ;
	break;
      case 'o':
	vp->type = OBJID;
	break;
      case 't':
	vp->type = TIMETICKS;
	break;
      case 'a':
	vp->type = IPADDRESS;
	break;
      default:
	fprintf(stderr, "bad type \"%c\", use \"i\", \"s\", \"x\", \"d\", \"n\", \"o\", \"t\", or \"a\".\n", *buf);
	return -1;
    }
    printf("Please enter the value: "); fflush(stdout);
    if (gets(buf) == NULL)
	exit(0);
    switch(vp->type){
      case INTEGER:
	vp->val.integer = (long *)malloc(sizeof(long));
	*(vp->val.integer) = atoi(buf);
	vp->val_len = sizeof(long);
	break;
      case STRING:
	if (ch == 'd'){
	    vp->val_len = ascii_to_binary(buf, value);
	} else if (ch == 's'){
	    strcpy(value, buf);
	    vp->val_len = strlen(buf);
	} else if (ch == 'x'){
	    vp->val_len = hex_to_binary(buf, value);
	}
	vp->val.string = (u_char *)malloc(vp->val_len);
	bcopy((char *)value, (char *)vp->val.string, vp->val_len);
	break;
      case NULLOBJ:
	vp->val_len = 0;
	vp->val.string = NULL;
	break;
      case OBJID:
	vp->val_len = MAX_NAME_LEN;;
	read_objid(buf, value, &vp->val_len);
	vp->val_len *= sizeof(oid);
	vp->val.objid = (oid *)malloc(vp->val_len);
	bcopy((char *)value, (char *)vp->val.objid, vp->val_len);
	break;
      case TIMETICKS:
	vp->val.integer = (long *)malloc(sizeof(long));
	*(vp->val.integer) = atoi(buf);
	vp->val_len = sizeof(long);
	break;
      case IPADDRESS:
	vp->val.integer = (long *)malloc(sizeof(long));
	*(vp->val.integer) = inet_addr(buf);
	vp->val_len = sizeof(long);
	break;
      default:
	fprintf(stderr, "Internal error\n");
	break;
    }

    return 1;
}

